### PR TITLE
Feat/stanza identification

### PR DESCRIPTION
**Pull request by alvp**
Added new stanza types from the stanza deliverable:
* seguidilla
* seguidilla_compuesta
* chamberga
* seguidilla_gitana
* endecha_real
* cuarteto_lira
* estrofa_sáfica  # se puede encadenar?
* estrofa_francisco_de_la_torre
* francisco_de_la_torre
* estrofa_manriqueña
* sexteto_lira
* septeto_lira
* ovillejo
* sonnet
* couplet
* tercetillo
* terceto
* sexteto
* sexteto_rima
* sextilla
* terceto_monorrimo
* redondilla
* aleluya
* cuarteto
* serventesio
* cuaderna_vía
* cuarteta
* octava_real
* copla_arte_mayor
* copla_arte_menor
* copla_castellana
* copla_mixta
* octava
* octavilla
* espinela
* copla_real
* lira
* quinteto
* quintilla
* silva_arromanzada
* cantar
* romance
* romance_arte_mayor
* haiku
* soleá
* decima_antigua
* septeto
* septilla
* novena

### CODE REVIEW

File: tests/test_rhymes.py [hunk 1]
```diff
@@ -470,3 +490,605 @@ def test_analyze_rhyme_romance(romance):
...
+
+
+def test_seguidilla():
+    poem = """Que se caiga la torre
+    de Valladolid
+    como a mí no me coja,
+    ¿qué se me da a mí?"""
+    output = "seguidilla"
+    input_poem = get_scansion(poem, rhyme_analysis=True)
+    assert input_poem[0]["structure"] == output
```
Comment (versae) [inline tests/test_rhymes.py:757]: Maybe put all the specific structure tests in its own test file.

File: tests/test_rhymes.py [hunk 2]
```diff
@@ -470,3 +490,605 @@ def test_analyze_rhyme_romance(romance):
...
+#     puede ningún amador
+#     es membrarse del placer
+#     en el tiempo del dolor;
+#     e ya sea que el ardor
+#     del fuego nos atormenta,
+#     mayor dolor nos aumenta
+#     esta tristeza y langor."""
+#     output = "octavilla"
+#     input_poem = get_scansion(poem, rhyme_analysis=True)
+#     assert input_poem[0]["structure"] == output
```
Comment (versae) [inline tests/test_rhymes.py:1022]: Remove?

File: src/rantanplan/rhymes.py [hunk 3]
```diff
@@ -32,6 +41,100 @@
 # will be chosen.
 STRUCTURES = (
```
[the remark this answers is not in the document]
Comment (alvp) [reply src/rantanplan/rhymes.py:42]: It's done

File: src/rantanplan/rhymes.py [hunk 4]
```diff
@@ -109,14 +384,32 @@ def get_clean_codes(stressed_endings, assonance=False, relaxation=False):
...
+            span = match.span()
+            syllable = (syllable[:span[0]] + match.group().upper()
+                        + syllable[span[1]:])
+        # Otherwise, only the final if there is a diphthong
+        else:
+            syllable = syllable.upper()
+        stressed_ending[stressed_position] = syllable
         # TODO: Other forms of relaxation should be tried iteratively, such as
         # lava ~ naba, vaya ~ valla, ceceo ~ zezeo, Venus ~ menos,
         # (also cases changing `i` for `e`), etc.
```
[the remark this answers is not in the document]
Comment (alvp) [reply src/rantanplan/rhymes.py:400]: Done

File: src/rantanplan/rhymes.py [hunk 5]
```diff
@@ -154,6 +447,7 @@ def assign_letter_codes(codes, code_numbers, unrhymed_verses, offset=None):
         if rhyme in unrhymed_verses:
             rhyme_letter = -1  # unrhymed verse
             endings.append('')  # do not track unrhymed verse endings
+            rhymes.append(rhyme_letter)
```
[the remark this answers is not in the document]
Comment (alvp) [reply src/rantanplan/rhymes.py:450]: Props to you!

File: src/rantanplan/rhymes.py [hunk 6]
```diff
@@ -281,3 +578,334 @@ def analyze_rhyme(lines, offset=4):
                 }
     if best_structure is not None:
         return best_structure
+
+
+def most_common_lengths(lengths_list, n=4, number_of_verses=None):
```
[the remark this answers is not in the document]
Comment (alvp) [reply src/rantanplan/rhymes.py:583]: Done, I think `n` is fine now it's explained in the docstring

File: src/rantanplan/rhymes.py [hunk 7]
```diff
@@ -281,3 +578,334 @@ def analyze_rhyme(lines, offset=4):
...
+    most_common = []
+    c = Counter(lengths_list).most_common(n)
+    if len(lengths_list) == number_of_verses or number_of_verses is None:
+        for length in c:
+            most_common.append(length[0])
```
[the remark this answers is not in the document]
Comment (alvp) [reply src/rantanplan/rhymes.py:588]: Added comment for that

File: src/rantanplan/rhymes.py [hunk 8]
```diff
@@ -281,3 +578,334 @@ def analyze_rhyme(lines, offset=4):
...
+    if number_of_verses == 4:
+        for i in (0, 2):
+            if 9 > lengths[i] > 5:
+                correct_lines += 1
+        for i in (1, 3):
+            if 7 > lengths[i] > 3:
+                correct_lines += 1
+    if correct_lines == number_of_verses:
+        return True
+    return False
```
[the remark this answers is not in the document]
Comment (alvp) [reply src/rantanplan/rhymes.py:629]: Done

File: src/rantanplan/rhymes.py [hunk 9]
```diff
@@ -281,3 +578,334 @@ def analyze_rhyme(lines, offset=4):
...
+    number_of_verses = len(lengths)
+    if number_of_verses == 4:
+        for i in (0, 1, 3):
+            if 8 > lengths[i] > 4:
+                correct_lines += 1
+        if 13 > lengths[2] > 9:
+            correct_lines += 1
+    if correct_lines == number_of_verses:
+        return True
+    return False
```
[the remark this answers is not in the document]
Comment (alvp) [reply src/rantanplan/rhymes.py:671]: I'll add an issue for that and work on this later on

File: tests/test_rhymes.py (hunk 1 above)
Comment (alvp) [reply tests/test_rhymes.py:757]: I'll add an issue for that and work on this later on

File: tests/test_rhymes.py (hunk 2 above)
Comment (alvp) [reply tests/test_rhymes.py:1022]: Yeah it's one of the stanzas I was having problems with, will remove and add later when it works

File: src/rantanplan/rhymes.py [hunk 10]
```diff
@@ -32,6 +41,100 @@
 # will be chosen.
 STRUCTURES = (
     (
+        CONSONANT_RHYME,
+        "seguidilla",
+        r"(-a-a)|(abab)",
+        lambda lengths: is_seguidilla(lengths)
```
[the remark this answers is not in the document]
Comment (alvp) [reply src/rantanplan/rhymes.py:47]: Done

File: src/rantanplan/structures.py [hunk 11]
```diff
@@ -0,0 +1,659 @@
...
+    """Return the n most commons lengths in a list of verses lengths
+
+    :param lengths_list: List with the lengths to be checked
+    :param n: Number of most common lengths to be returned
+    :param number_of_verses: How many verses are in a given stanza
+    :return: The n most common lengths
+    """
+    most_common_lengths_list = []
+    count = Counter(lengths_list).most_common(n)
+    if len(lengths_list) == number_of_verses or number_of_verses is None:
```
Comment (versae) [inline src/rantanplan/structures.py:26]: The line `count = Counter(lengths_list).most_common(n)` can be placed inside the `if` statement.

File: src/rantanplan/structures.py [hunk 12]
```diff
@@ -0,0 +1,659 @@
...
+        for length in count:
+            # `lengths` is a tuple with the length and it's count so we only get
+            # the first element of the tuple
+            most_common_lengths_list.append(length[0])
```
Comment (versae) [inline src/rantanplan/structures.py:30]: I see. Thanks for the clarification. Then, maybe something like next would help
```
for length, _ in count:
    most_common_lengths_list.append(length)
```
Up to you.

File: src/rantanplan/structures.py [hunk 13]
```diff
@@ -0,0 +1,659 @@
...
+    if number_of_verses == 4:
+        for i in (0, 2):
+            if 9 > lengths[i] > 5:
+                correct_lines += 1
+        for i in (1, 3):
+            if 7 > lengths[i] > 3:
+                correct_lines += 1
+    if correct_lines == number_of_verses:
+        return True
+    return False
```
Comment (versae) [inline src/rantanplan/structures.py:69]: Replace with `return correct_lines == number_of_verses` in all functions where applicable.

File: src/rantanplan/structures.py (hunk 11 above)
Comment (alvp) [reply src/rantanplan/structures.py:26]: Done

File: src/rantanplan/structures.py (hunk 12 above)
Comment (alvp) [reply src/rantanplan/structures.py:30]: Yes good idea, better  that way

**Pull request merged**